File: Introductory Problems/cpp/TowerofHanoi.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long

void towerOfHanoi(ll n, char source, char middle, char destination) {
    // base case
    if (n == 1) {
        cout << source << " " << destination << '\n';
        return;
    }
    towerOfHanoi(n-1, source, destination, middle);
    cout << source << " " << destination << '\n';
    towerOfHanoi(n-1, middle, source, destination);
}

// main function
int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);
 
    ll n;
    // no. of disks
    cin >> n;
 
    // tower of hanoi has 2^n - 1 steps
    cout << (1<<n)-1 << "\n";
    // solve
    towerOfHanoi(n, '1', '2', '3');
    return 0;
}

File: Introductory Problems/java/TowerofHanoi.java
import java.util.Scanner;

public class TowerofHanoi {

    private static void towerOfHanoi(int n, char source, char middle, char destination) {
        // base case
        if (n == 1) {
            System.out.println(source + " " + destination);
            return;
        }

        towerOfHanoi(n - 1, source, destination, middle);
        System.out.println(source + " " + destination);
        towerOfHanoi(n - 1, middle, source, destination);
        
        return;
    }

    public static void main(String[] args) {  
        // scanner
        Scanner sc = new Scanner(System.in);
        
        // no. of disks
        int n = sc.nextInt();
        // closing input stream
        sc.close();

        // tower of hanoi has 2^n - 1 steps
        System.out.println((1<<n)-1);
        
        // solve
        towerOfHanoi(n, '1', '2', '3');

        return;
    }    
}
